### PR TITLE
Add weback-dev-server with hot/hmr support

### PR DESCRIPTION
**Pull request by davkal**
* adds `npm start` / `yarn start` script
* starts a webpack-dev-server using the dev config, served on :3333
* all non-`public/build` requests proxied to Grafana server on :3000
* webpack dev config now clears `public/build`
* hot reloading (HMR) for react/styles, not working for angular code
* new entry `dev.ts` for dynamic imports of CSS theme (ExtractText does
not work with HMR)
* TS loader pipeline moved out of common to add HMR for react
* applied `hot()` to some react containers (that's their new default
 export, named exports remains for testing)
* added sections to README
* updated yarn.lock
* moved `loadStore()` calls in react component constructors to `componentDidMount` since they have sideeffects

Minimum test plan:
- [ ] `yarn start` and open on localhost:3333 and change a tsx files and see it hot replaced
- [ ] `yarn watch` and open on localhost:3000 make sure old dev flow still works
- [ ] `make build` serve all from binary

### CODE REVIEW

**Comment by davkal** [comment]
I know why *test-frontend* failed, the things I moved out of webpack common need to be in the test config too. Fix coming up...

File: public/app/containers/AlertRuleList/AlertRuleList.tsx [hunk 1]
```diff
@@ -173,3 +174,5 @@ export class AlertRuleItem extends React.Component<AlertRuleItemProps, any> {
     );
   }
 }
+
+export default hot(module)(AlertRuleList);
```
Comment (torkelo) [inline public/app/containers/AlertRuleList/AlertRuleList.tsx]: Having to wrap all containers like this seems a bit intrusive, and repetitive. can’t this be done centrally, for example in react container (the angular component that loads the page react component)

**Comment by davkal** [comment]
I've looked at centralized alternatives, but did not succeed:

* `module.hot.accept()` needs the import path to know what module to reload
* `hot(module)($route.current.locals.component)` won't work because it's not the component's `module` that's being referenced

In `ReactContainer` we don't have access to the component's module nor the import path.

My suggestion is to move ahead with the individual `hot()` calls. I think it will be less of a problem in the future because devs will look at/clone existing classes when writing more components.

**Comment by torkelo** [comment]
Dam, just looks so ugly having to include that dev tool in all containers. Looking at a the example in the react hot reloader readme they all just wrap the app component. But if there is no other way...

**Comment by torkelo** [comment]
This seems to also mess up the reuse of rules that was in common file duplicating it in dev and test and prod :( 

Also this might slow down the build a lot since sass source mapp is enabled, disabling that had significant speed boost. But not sure since I have not tested this, but if you notice a slow down with his please diasable the source mapp as it’s rarely used and a 2-3 second build time improvement is more important :)

File: scripts/webpack/sass.rule.js [hunk 2]
```diff
@@ -23,8 +23,14 @@ module.exports = function(options) {
             config: { path: __dirname + '/postcss.config.js' }
           }
         },
-        { loader:'sass-loader', options: { sourceMap: options.sourceMap } }
+        { loader: 'sass-loader', options: { sourceMap: options.sourceMap } }
       ],
+      fallback: [{
+        loader: 'style-loader',
```
Comment (torkelo) [inline scripts/webpack/sass.rule.js]: I think we can just use the ExtractTextPlugin (and skip the extractSass bool logic) and just add a fallback option to the ExtractTextPlugin

**Comment by torkelo** [comment]
It seems like source maps with sass is much faster without the extract text plugin, so adding it back does not slow it down that bad (300ms) while using hot moduke reload (since it is not using the extract text plugin). 

It seems we are also piping typescript through babel with this change, which seems to add ~500ms or so. Is this really needed? 

Running the non hot module reload (npm run watch) it seems to have slowed down by ~3 seconds, mostly due to sass source map being enabled. 

Otherwise seems to work pretty well, will be intesting how useful and stable this will be. My previous experiance with HMR has been that is nice (especially for css work) but buggy and slow so can be frustrating as well.

**Comment by torkelo** [comment]
Btw, sorry if I sound negative with these comments. Think it’s great that you got this to work, did not think it would be possible in mixed angular + react app like grafana :) just a bit worried about the complicating and slowing down the build and making dev builds more different that production builds than they already are. 

But the productivity gain here if it works correctly will probably make it worthwhile.

**Comment by davkal** [comment]
> This seems to also mess up the reuse of rules that was in common file duplicating it in dev and test and prod :(

When it comes to config I'm in favour of dropping reuse when it needs added logic, since IMO config needs to be as explicit as possible. Dev needed an additional loader in the pipeline. I'm happy to rewrite this with list logic.


> Also this might slow down the build a lot since sass source mapp is enabled, disabling that had significant speed boost. 

I'm currently relying on this to find the origin on styles. Although I have some trouble viewing the maps. Will revisit its usefulness.

> It seems we are also piping typescript through babel with this change, which seems to add ~500ms or so. Is this really needed?

The react-hot-loader is a babel plugin, hence we needed to add babel to the pipeline (but only in dev).

**Comment by torkelo** [comment]
this PR broke having Grafana behind subpath (via nginx reverse proxy) like http://myserver/grafana, fixed by this PR https://github.com/grafana/grafana/pull/11804 which also required reverting the change done here with the preserveUrl parameter. Had to restore that to false for fonts/images to load.

**Pull request merged**